### DISCUSSION TITLE
General Chat

### DISCUSSION THREAD

**Discussion by AndyPro720** [Unified-Development-Environment/Unified_HAL#1] [General]
@rish292 We can talk about backend here, so that it is tabeled

**Comment by AndyPro720** [comment]
I'll correct and push on arduino_dev branch.

**Reply by rish292** [reply]
Okay

**Reply by AndyPro720** [reply]
@rish292  So as for PRs we can create them ever so occasionally when we need push changes to main branch. Later we can recreate the branch with same name, so the local git branches remain unchanged, (naming is key). 

Baki on when to merge, I think it's best when one abstraction is finished.

**Reply by rish292** [reply]
Hmm

**Reply by AndyPro720** [reply]
@rish292 So we have the exact function across all systems to be like so
`void uf_gpioWrite(char PORT[] , int PIN , int VALUE)`

**Reply by AndyPro720** [reply]
And the conversions will be done in function definition for the respective vendor

**Reply by AndyPro720** [reply]
Updated in wiki as well

**Comment by AndyPro720** [comment]
Hey @rish292 , I was wondering

**Reply by AndyPro720** [reply]
We thought of having a common header file that selects the microcontroller file depending on a certain definition, which means our package will have to have the selector header file -> the HAL header file -> the c file. Plus while we have one common selector header, the package will need to contain `all implementations and headers`, making it heavy, even though we'll be leveraging only one or two.  

Plus, there's an extra step involved where the linker needs to work its way through the selector header file and onto the HAL header

**Reply by AndyPro720** [reply]
The other approach, or the one that we seem to have adopted subconsciously is to have each vendor MCU packaged as a separate library and the user selects the package he wants and installs it to make the `main` code compatible with another system. Less overhead, needs a ui implementation, perhaps the extension lib we talked about earlier.

One problem that could occur though is that the two packages could cause conflict if included in the same env. Like say the STMm4 and STMm7, where the #include will be the same for both.

**Reply by AndyPro720** [reply]
So here we either leave it on the user, but that seems like an intentionally bad design choice, one that could even take a headache to debug.

* Another option could be to introduce a selector header for the same vendor. So for each STM package we include that one file.
* But here again, NodeMcu is a different vendor but can be programmed in Arduino IDE, so could again cause conflicts here.
* Or even better, as I write, what if we have a universally common selector header file, included say with each package, but the header and implementation is only distributed in the specific vendor package.
* We could also rename the header as `uf_mcu` directly to avoid this mess entirely. It does mean the header file name won't be common for each mcu. Seems like a fair trade off.

**Reply by AndyPro720** [reply]
~Or if you want to reduce the packages the maybe `uf_vendor`, but we'll need a selector header then.  But thing is they'll still have to #define the mcu in either case.~  So better keep the header file to do that.

I'm trying to find a benefit to the selector header file in any case over the uf_mcu method. I can't think of one, how about you?

**Reply by AndyPro720** [reply]
Assuming we go with `uf_mcu` method, we'll have two packages I suppose

* one ui based where they download there specific ide and mcu package. Can be imported into the env. Just like we have been doing.
* The other is a universal package, that has all headers and implementations, and can be used in generic development by #include statements. (ofc assuming the generic env doesn't need any sub directories and specific format.


//Also; we need to work out the makefile thing as well

**Reply by rish292** [reply]
I think we should create packages which would mesh with the ides we are using so it would be similar to the first point. As far as the makefile, that linking problem isn't arising with the CubeIDE and Keil. It's only the Arduino. So the end user would need these softwares in order to use our lib.
@AndyPro720

**Reply by AndyPro720** [reply]
@rish292 Didn't you mention if the files are put directly from outside and not created within the env they don't get detected. Or you'll need a very specific format, that you showed earlier. Makefile should solve that. 
Arduino has a well documented process for its libs so no hassle importing and running code there, have been and infact had pushed a packaged library with keyword highlighting a day or two back.

**Reply by rish292** [reply]
Nope I have tried the manual add method with another test project I created. I added the .h and .c files from our Unified HAL folder onto the project's src and inc folder and they're working fine. @AndyPro720

**Comment by AndyPro720** [comment]
https://beej.us/guide/bgc/html/split-wide/the-c-preprocessor.html#the-c-preprocessor

**Reply by AndyPro720** [reply]
@rish292
